Add inpainting arrow

russelldj opened the pull request:
Add functionality to do Telea, Navier-Stokes, and black mask inpainting. Also update the burnout pipeline to use this by default.

mleotta commented on arrows/ocv/inpaint.h:
```diff
@@ -0,0 +1,55 @@
+  const std::unique_ptr< priv > d;
+};
+
+} // namespace vxl
```
should be `ocv`

mwoehlke-kitware commented on config/pipelines/detect_burnin.pipe:
```diff
@@ -75,12 +75,23 @@ process detect_burnin__dilater
 
 # =============================================================================
-process detect_burnin__out
+process detect_burnin__convert
```
So... this should probably be named  `detect_burnin__mask_out`, and the other `detect_burnin__inpainted_out`, or something, so that it's more obvious how to get both outputs. (The GUI needs both!)

russelldj replied:
Updated to `detect_burnin__mask`. If you feel strongly that the `_out` postfix is useful, I'll add it in a couple of other places as well.

mwoehlke-kitware replied:
I would actually leave it `out`. The reason is that `out` is... the output of the "superprocess". Similarly, the new `.pipe` should have an `in` and an `out`, even if that requires creating a "dummy" process. (Which it looks like it will...)

mleotta replied:
This may be one of the key difference between "clusters" and including pipe files.  In a cluster we define the inputs and outputs of the cluster as if it was just a process, but these ports are really just and alias to ports on internal nodes, so there is no "dummy" process created.  The cluster defines what inputs and outputs are accessible and hide the rest.  When we are just including a pipe file that whole pipeline is exposed and anyone can tap into any ports they want.  It might be that for our current applications the outputs are the mask and inpainted results, but these have no special status.  Someone can just as easily tap into any port of the included file. 

I think it's better for the processes to be named in a way that documents what they do rather that how we anticipate this pipe file might be used by someone who includes it.

mwoehlke-kitware replied:
The problem with that approach is that it requires the user to have knowledge of the internal implementation of the `.pipe`. That's fine if you *want* to inspect the guts, but if you just want to use the thing, it makes it more complicated and less obvious what you're trying to do at the point of use.

That said, I think what would be useful in the long term is some sort of hybrid, where we can define a "cluster" that allows poking at the guts (and that follows the config search paths rather than having its own, independent system), but also allows some sort of scoping of its internals with the ability to specify "outward-facing" ports.

russelldj replied:
I see the appeal of simply naming it `out` but I also think that it's a bit weird to give most things a descriptive name and then have the only thing the user touches have no semantic meaning. Furthermore, unless we move toward better documenting things, you'd still need to inspect an embedded pipeline to determine the types of the ports you're getting, so it seems reasonable to figure out the name as well. `<super process>__<semantic>_out` seems like a reasonable convention if we want to clue users in to things which may be useful. But I think `<super process>__<semantic>` is nearly as good and might be cleaner if there are multiple optional outputs. 

I'll look into creating a dummy process for the inpainter input.

mwoehlke-kitware commented on arrows/ocv/register_algorithms.cxx:
```diff
@@ -37,6 +37,7 @@
 #include <arrows/ocv/feature_detect_extract_ORB.h>
 #include <arrows/ocv/feature_detect_extract_SIFT.h>
 #include <arrows/ocv/feature_detect_extract_SURF.h>
+#include <arrows/ocv/inpaint.h>
 #include <arrows/ocv/image_io.h>
```
Pedantic: `'n'` > `'m'`:
```suggestion
#include <arrows/ocv/image_io.h>
#include <arrows/ocv/inpaint.h>
```

(I wonder if the registrations should be in some sort of order, also? Right now it seems they aren't.)

russelldj replied:
It seems like they're loosely grouped by functionality but yes, I agree this isn't great.

mwoehlke-kitware replied:
I wouldn't be surprised if they're in the order the algorithms were created 🙂. The second comment stems mainly from noting that `inpaint` and `image_io` were in that order for registration, also, but then followed by something that was definitely *not* lexicographical order.

mwoehlke-kitware commented on arrows/ocv/inpaint.h:
```diff
@@ -0,0 +1,55 @@
+private:
+  class priv;
+
+  const std::unique_ptr< priv > d;
```
```suggestion
  std::unique_ptr< priv > const d;
```

mwoehlke-kitware commented on arrows/ocv/inpaint.h:
```diff
@@ -0,0 +1,55 @@
+
+  /// Get this algorithm's \link vital::config_block configuration block
+  /// \endlink
+  virtual vital::config_block_sptr get_configuration() const;
```
I think these can be marked `override`?
```suggestion
  vital::config_block_sptr get_configuration() const override;
```

mwoehlke-kitware commented on arrows/ocv/inpaint.h:
```diff
@@ -0,0 +1,55 @@
+{
+public:
+  PLUGIN_INFO( "ocv_inpainting",
+               "Inpaint a RGBA image using non-zero A values as a mask." )
```
```suggestion
               "Inpaint an RGBA image using non-zero alpha values as a mask." )
```
(I could probably go either way on 'A' vs. 'alpha', but "**an** RGBA", please 🙂 .)

mwoehlke-kitware commented on arrows/ocv/CMakeLists.txt:
```diff
@@ -35,6 +35,7 @@ set( ocv_headers_public
   feature_detect_extract_SURF.h
   detect_heat_map.h
   hough_circle_detector.h
+  inpaint.h
   image_container.h
```
```suggestion
  image_container.h
  inpaint.h
```

mwoehlke-kitware commented on arrows/ocv/CMakeLists.txt:
```diff
@@ -88,6 +89,7 @@ set( ocv_sources
   feature_detect_extract_ORB.cxx
   feature_detect_extract_SIFT.cxx
   feature_detect_extract_SURF.cxx
+  inpaint.cxx
   image_container.cxx
```
```suggestion
  image_container.cxx
  inpaint.cxx
```

(Given the consistency, I'm curious if this name changed at some point while you were working on this? 😉 )

russelldj replied:
I think I was just consistently bad at the alphabet.

mwoehlke-kitware commented on arrows/ocv/inpaint.cxx:
```diff
@@ -0,0 +1,169 @@
+  METHOD_navier_stokes,
+};
+
+ENUM_CONVERTER( method_converter, inpainting_method, { "mask", METHOD_mask },
+                { "telea", METHOD_telea },
```
This is definitely an "up to you", but maybe...
```suggestion
ENUM_CONVERTER( method_converter, inpainting_method,
                { "mask", METHOD_mask },
                { "telea", METHOD_telea },
```

mwoehlke-kitware commented on arrows/ocv/inpaint.cxx:
```diff
@@ -0,0 +1,169 @@
+class inpaint::priv
+{
+public:
+
```
```suggestion
```

russelldj replied:
This is an uncrustify regression.

mleotta commented on config/pipelines/detect_burnin.pipe:
```diff
@@ -75,12 +75,23 @@ process detect_burnin__dilater
+  merge_images:type                     = ocv
+
+# =============================================================================
+process detect_burnin__out
```
I think maybe we should call this `inpaint` rather than `out`.  There may be multiple "outs" if the user of this file wants to save both masks and inpainted results.

mwoehlke-kitware commented on arrows/ocv/inpaint.cxx:
```diff
@@ -0,0 +1,169 @@
+  {
+  }
+
+  ~priv()
+  {
+  }
+
```
```suggestion
```

mwoehlke-kitware commented on arrows/ocv/inpaint.cxx:
```diff
@@ -0,0 +1,169 @@
+  // get base config from base class
+  vital::config_block_sptr config = algorithm::get_configuration();
+
+  config->set_value( "inpaint_method", method_converter().to_string(
+                       d->method ),
```
```suggestion
  config->set_value( "inpaint_method",
                     method_converter().to_string( d->method ),
```

mwoehlke-kitware commented on arrows/ocv/inpaint.cxx:
```diff
@@ -0,0 +1,169 @@
+inpaint
+::get_configuration() const
+{
+  // get base config from base class
```
```suggestion
  // Get base config from base class
```

mwoehlke-kitware commented on arrows/ocv/inpaint.cxx:
```diff
@@ -0,0 +1,169 @@
+  if( radius <= 0 )
+  {
+    LOG_ERROR( logger(),
+               "Radius should be positive but instead was " << radius );
```
```suggestion
               "Radius should be non-negative but instead was " << radius );
```
...oh, and it occurs to me, is it possible to feed in a non-finite value? I believe that would be accepted by this check, though I'm not sure the config parser accepts non-finite values.

russelldj replied:
0 is not an acceptable value as this is written, so I believe that the original message is correct. I'm not sure if it's possible to have an infinite value in a config, but I feel like you'd have to try pretty hard to mess up that badly. My intuition is anything like `inf` would get parsed as a string or fail.

mwoehlke-kitware replied:
Ack, somehow I was thinking this was written in a way that would accept zero :flushed:.

Nothing to see, here. Move along.

mleotta commented on config/pipelines/detect_burnin.pipe:
```diff
@@ -75,12 +75,23 @@ process detect_burnin__dilater
   filter:vxl_convert_image:scale_factor = 255
 
+# =============================================================================
+process detect_burnin__merge
```
I'm wondering if it makes sense to have a separate `inpaint_burnin.pipe` that includes `detect_burnin.pipe` and adds the inpainting parts.  Some users may want to use just the detector.  Another option is to have `inpaint_burnin.pipe` include just these two process and not include `detect_burnin.pipe` and then require the user to include both and connect them up.  I'm not sure which is better.

mwoehlke-kitware replied:
Actually, that's probably a better idea... or even, assumes that whoever includes it has a `detect_burnin__out`. (Also, while I didn't think of it earlier, I'm not sure inpainting belongs in "**detect**_burnin.pipe".)

russelldj replied:
I followed Matt's initial suggestion of making a `inpaint_burnin.pipe` which is a very light wrapper around `detect_burnin.pipe`. I made a small breaking change to the existing functionality which was updating `detect_burnin__out` to `detect_burnin__mask`.

mwoehlke-kitware commented on arrows/ocv/inpaint.cxx:
```diff
@@ -0,0 +1,169 @@
+    }
+    case METHOD_mask:
+    {
+      cv::Mat zeros = cv::Mat::zeros( cv_image.size(), cv_image.type() );
```
I guess it's not necessary to do now, but maybe as a follow-up, should we allow the user to configure the inpainting color?

russelldj replied:
I agree this could be useful later. If you have an example handy of specifying a multi-element config value, since I'd probably let the user specify an RGB tuple, that would be useful.

mleotta replied:
I'm sure we could find a cleaner way, but here is an example:
https://github.com/Kitware/kwiver/blob/master/arrows/cuda/integrate_depth_maps.cxx#L160

We just read/write the value as a string and then convert it with iostream

mleotta commented on arrows/ocv/inpaint.h:
```diff
@@ -0,0 +1,55 @@
+
+/// OCV image inpainting process
+///
+/// Create a masked RGB image from a RGBA image
```
This comment is not very clear.  You might want to write out more clearly what it means to inpaint and that the pixels from the alpha channels are used as a mask to indicate which pixel to paint.

mwoehlke-kitware reviewed (changes requested):
Looks pretty good. Have we considered, though, whether using the alpha channel is the right way to go about this, or if we should instead take multiple image inputs? The former makes size mismatches impossible and may be more convenient. The latter would allow operating on N-channel images and accepting the mask as a `bool` image.

mleotta reviewed (changes requested):
Looks pretty good.  I have a few minor comments.  It's probably also a good idea to add a simple test.  Maybe check that inpainting changes pixels in the mask but not outside it.

russelldj commented:
> Looks pretty good. Have we considered, though, whether using the alpha channel is the right way to go about this, or if we should instead take multiple image inputs? The former makes size mismatches impossible and may be more convenient. The latter would allow operating on N-channel images and accepting the mask as a `bool` image.

You're probably right that it's better to have this take two images. I can update this so it's an implementation of `merge_images`. Size checking should be easy as that logic is already in the merger.

russelldj commented:
> Looks pretty good. I have a few minor comments. It's probably also a good idea to add a simple test. Maybe check that inpainting changes pixels in the mask but not outside it.

For other filters I've been pretty happy running the filter once, validating the result, and then ensuring that future tests produce the exact same result. I'm open to writing a more specialized test for this application but don't currently see any benefit to that.

mwoehlke-kitware commented on arrows/ocv/inpaint.h:
```diff
@@ -0,0 +1,55 @@
+
+namespace ocv {
+
+/// OCV image inpainting process
```
```suggestion
/// OCV image inpainting process.
```

mwoehlke-kitware commented:
> I can update this so it's an implementation of `merge_images`.

Ah, yes, that probably makes sense. (I also wasn't thinking that, of course, we already have code to handle two input images.)

In that case, I'm not sure it's even worth modifying `detect_burnin.pipe`.

> Size checking should be easy as that logic is already in the merger.

Right. I didn't mean to imply otherwise, rather I was saying that with a single input, you *can't* mess that up. That is perhaps a rather tenuous "advantage" though.

russelldj commented:
I think all relevant comments have been addressed.

russelldj commented:
I'm a bit stumped by the issue I brought up in #1279 about failing to register the dummy process. I'm happy to get this in and do the pipeline updates on a different PR or if there's no rush, waiting until that's addressed.

russelldj commented:
Based on Linus' feedback, I fixed the issue and added a dummy input to the inpainter. To your comment in that issue Matthew, I don't see a reason to have a dummy output since the inpainter only provides one output. I set it up so you connect to both the input and output of the inpainter process. What I'm guessing you were describing was a situation like the other one Matt described, with a separate detector and inpainter the user has to connect. However, that seems rather unhelpful to me since the inpainter is only a single process.

mwoehlke-kitware commented on arrows/ocv/inpaint.cxx:
```diff
@@ -0,0 +1,165 @@
+  {
+    case METHOD_telea:
+    {
+      cv::inpaint( cv_image, cv_mask, cv_image, d->radius,
```
Heh, this seems suspiciously more convenient now vs. the old way of extracting a channel as the mask 🙂.

mwoehlke-kitware commented on arrows/ocv/inpaint.cxx:
```diff
@@ -0,0 +1,165 @@
+  {
+    LOG_ERROR(
+      logger(),
+      "Image and masks sizes were different. " << cv_image.size() << " vs. " <<
+        cv_mask.size() );
```
Not important, but is this used elsewhere? I probably would have written something closer to "Image size {} does not match mask size {}", but if similar text is used elsewhere, we should probably leave it alone. (OTOH, if we're using a different message elsewhere, being consistent would probably be good.)

russelldj replied:
I didn't specifically try to mimic another pattern here. I'll update it.

mwoehlke-kitware commented on arrows/ocv/tests/test_inpaint.cxx:
```diff
@@ -0,0 +1,108 @@
+  auto const inpainted_image_ptr = inpainter.merge( input_image, input_mask );
+
+  auto expected_filename = data_dir + "/" + expected_basename;
+  auto const expected_image_ptr = io.load( expected_filename );
```
Would it be worth asserting that the images loaded successfully?

russelldj replied:
It seems like the traceback would be fairly self explanatory if image loading failed. I'm open to adding this check but it seems like a better solution might be to add an `ocv::image_io` unit test as done in `vxl`.

mwoehlke-kitware replied:
I'm not especially worried about the traceback, as we aren't testing image I/O. I'm more worried about the test blowing up in unnecessarily spectacular fashion (e.g. SEGV, which it looks like *will* happen) if, say, one of the images passed to `inpainter.merge` is a null pointer. But I'm also not sure it's even worth worrying that the testing environment has gone sideways, which is why the original comment is a question rather than a suggestion. (Note: if I was going to do something, I would probably `ASSERT_NE(nullptr, input_image)`, and so forth.)

To be clear, I don't particularly object to `inpainter.merge` crashing if you give it null pointers. I rather doubt we tend to check that sort of thing. Don't do that. Usually, we expect the *user* to ensure they don't do that. In this case, the user is the unit test, so one might argue for a paranoia check that we actually did load the images. But, again, something is *really* wrong if that happen, possibly to the extent that "it broke; here's the pieces" is a satisfactory reaction.

russelldj replied:
I don't have strong feelings any way here. However, if we make a decisions to validate image loading there's 10ish `vxl` tests that need to be updated. Which makes be wonder if this is better served as a separate discussion.

mwoehlke-kitware commented on config/pipelines/inpaint_burnin.pipe:
```diff
@@ -0,0 +1,23 @@
+connect from inpaint_burnin__in_image.image
+        to   detect_burnin__in_image.image
+
+connect from detect_burnin__in_image.image
```
```suggestion
connect from inpaint_burnin__in_image.image
```

mwoehlke-kitware reviewed (changes requested):
(-1 is just for the `.pipe` comment...)

mleotta commented:
Jenkins test this please

russelldj commented:
I'm a bit lost what's going on here. Both Mac and Windows are failing on `292 - ocv:inpaint.telea`, though the tests for other inpainting strategies are passing. The issue is reported to be at the comparison between the computed and expected results. I've run hundreds of iterations of the test locally on Linux without issue so I don't think this is about determinism. 

@dstoup @Cookt2 have you seen discrepancies in OpenCV computations between OSs? Is there any workflow for diagnosing failing tests other than getting access to the appropriate OS? If it's not an issue with computation, it's conceivable that the expected image for just that test is being read improperly, but that seems very unlikely. However, maybe a test for `ocv` IO is useful, as I added for `vxl`

A quick and dirty solution would be to remove this test and stick to using one of the other methods, `navier_stokes` or `mask`.

russelldj commented:
For the sake of getting this merged I removed the Telea method from both the tests and the code.

dstoup commented:
> I'm a bit lost what's going on here. Both Mac and Windows are failing on `292 - ocv:inpaint.telea`, though the tests for other inpainting strategies are passing. The issue is reported to be at the comparison between the computed and expected results. I've run 1000 iterations of the test locally on Linux without issue so I don't think this is about determinism.
> 
> @dstoup @Cookt2 have you seen discrepancies in OpenCV computations between OSs? Is there any workflow for diagnosing failing tests other than getting access to the appropriate OS? If it's not an issue with computation, it's conceivable that the expected image for just that test is being read improperly, but that seems very unlikely. However, maybe a test for `ocv` IO is useful, as I added for `vxl`
> 
> A quick and dirty solution would be to remove this test and stick to using one of the other methods, `navier_stokes` or `mask`.

I don't know about OpenCV specifically but I have seen differences in numerical computations on different platforms. Usually they do crop up in places like comparing computed vs. expected results.

Some of the things I've seen done in the past are:
- do the equivalent of TEST_NEAR instead of TEST_EQ in the case where close enough is good enough and if the differences are are small floating point discrepancies.
- In vidtk recently, we used a seed to help with consistency in the case where we were generating random pixel values for normalization.

Unfortunately I can't think of a good way to diagnose these types of issues without having access to a machine where the failure happens. Since most of the CI machines are on the domain, you should be able to access them but it's tricky if you want to access the PR builds directly. Since a PR build ends up overwriting the previous PR you would need to get on the machine after you PR fails and before the next one starts. Usually I run the PR at a time when there isn't much activity so it won't get clobbered, access the machine and copy the src/build directories out of the way. You need to change the paths in the CMakeCache.txt file to make it work, but that's one approach. Before doing that though, you might just try getting on one of them and building kwiver on your own, against the saved night Fletch build. Most failures are reproducible that way.

I would be happy to work with you on it if you need. Since you took the test out of this PR, I suggest maybe making a WIP PR that just puts it back, then we have a way to explore the failure in more depth.

mleotta commented:
If I recall, there is a way to specify image outputs for failing tests so that they show up in CDash and you can review them.

https://stackoverflow.com/questions/50719666/upload-image-diff-using-ctest-and-cdash/50760569

mwoehlke-kitware commented:
@mleotta, it appears this is waiting on you to re-review?

mleotta commented:
I'm not going to have time to review it.  @mwoehlke-kitware I'll trust your final review.

mwoehlke-kitware commented:
@mleotta, to be fair, I was referring specifically to clearing your "changes requested"... so as long as you're satisfied that the changes you requested were indeed made, that's fine 🙂.
